bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java

```diff
@@ -1426,14 +1426,23 @@ private static void execStoreOpcodes(WorkerExecutionContext ctx, WorkerData sf,
                 i = operands[0];
                 j = operands[1];
                 k = operands[2];
-                bArray = (BRefValueArray) sf.refRegs[i];
-                if (bArray == null) {
+                BNewArray list = (BNewArray) sf.refRegs[i];
+                if (list == null) {
                     handleNullRefError(ctx);
                     break;
                 }
 
                 try {
-                    bArray.add(sf.longRegs[j], sf.refRegs[k]);
+                    long index = sf.longRegs[j];
+                    BRefType refReg = sf.refRegs[k];
+                    BType elementType = (list.getType().getTag() == TypeTags.ARRAY_TAG)
+                            ? ((BArrayType) list.getType()).getElementType()
+                            : ((BTupleType) list.getType()).getTupleTypes().get((int) index);
+                    if (!checkCast(refReg, elementType)) {
+                        throw BLangExceptionHelper.getRuntimeException(RuntimeErrors.INCOMPATIBLE_TYPE,
```

Comment:
I think you should handle the error here (like set the error to ctx and handle error)